Antialiasing, EPS surfaces, and PNG performance

asinghvi17 opened the pull request:
This is the first stage of the refactor I've been working on.  Once this is merged into master, I can rebase the other branches I have onto this and make some manual edits.

asinghvi17 commented on src/CairoMakie.jl:
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
+# Default to ARGB Surface as backing device
+# TODO: integrate Gtk into this, so we can have an interactive display
+function CairoScreen(scene::Scene; device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
     w, h = size(scene)
```
```suggestion
    w, h = round.(Int, scene.camera.resolution[] .* device_scaling_factor)

asinghvi17 commented on src/CairoMakie.jl:
```diff
@@ -817,23 +815,34 @@ function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"applicati
+end
+
 function AbstractPlotting.backend_show(x::CairoBackend, io::IO, m::MIME"image/png", scene::Scene)
+    screen = CairoScreen(scene; device_scaling_factor = px_per_unit)
```
```suggestion

asinghvi17 commented:
CI failures are expected and confirm that everything works correctly.

jkrumbiegel commented on src/CairoMakie.jl:
```diff
@@ -817,23 +814,34 @@ function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"applicati
 
+
+function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"application/postscript", scene::Scene)
+    screen = CairoScreen(scene, io; mode=:eps)
```
isn't the pt_per_unit thing missing here?

asinghvi17 replied:
Yes, good point.  Will add that in.

jkrumbiegel commented on src/CairoMakie.jl:
```diff
@@ -372,17 +389,19 @@ function to_cairo_image(img::AbstractMatrix{<: AbstractFloat}, attributes)
+            for i in size(img, 2):-1:1, # account for Y-axis discrepancy in Cairo
+                j in 1:size(img, 1)
+        ]
+    )
```
is this the same as `collect(img')`? Maybe document more what's going on here. Probably the column major, row major conversion?

asinghvi17 replied:
This is just flipping the y-axis, it needs to be transposed for conversion to a surface.  I didn't change the code - just formatted it :D.   Will add docs.

jkrumbiegel commented on src/CairoMakie.jl:
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
 
-function CairoScreen(scene::Scene, path::Union{String, IO}; mode = :svg, device_scaling_factor = 1)
+function CairoScreen(scene::Scene, path::Union{String, IO}; mode = :svg, device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
 
```
I'm not sure that svg needs an additional antialias keyword, that seems like a png thing

jkrumbiegel commented on src/CairoMakie.jl:
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
 end
 
-function CairoScreen(scene::Scene, path::Union{String, IO}; mode = :svg, device_scaling_factor = 1)
+function CairoScreen(scene::Scene, path::Union{String, IO}; mode = :svg, device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
```
i would just remove the `mode = :svg` default and make it a required keyword. it's not obvious why svg would be a meaningful default and then you don't know what's happening in code that calls this

asinghvi17 commented:
Just rebased on master.  I think I addressed the comments you had in the commit comments, Julius (documentation of the image rotation and making the mode a required arg in `CairoScreen(scene, io, mode)`).

asinghvi17 commented on src/CairoMakie.jl:
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
+# Default to ARGB Surface as backing device
+# TODO: integrate Gtk into this, so we can have an interactive display
+function CairoScreen(scene::Scene; device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
+    w, h = round.(Int, scene.camera.resolution[] .* device_scaling_factor)
```
The advantage here is that you don't need to provide an unnecessary IO stream (which doesn't make sense for PNG surfaces).

asinghvi17 commented:
There is one test which fails because PlotUtils defines a colormap called `:grays`, and for some random reason, `to_color` can sometimes convert colors to colormaps.  Since it's an upstream issue, there's nothing we can do here.  Will make the merge, thanks Julius!
